libs/wire-api/src/Wire/API/User/RichInfo.hs

```diff
@@ -244,16 +248,17 @@ instance Monoid RichInfoAssocList where
 instance Semigroup RichInfoAssocList where
   RichInfoAssocList a <> RichInfoAssocList b = RichInfoAssocList $ a <> b
 
-instance ToJSON RichInfoAssocList where
-  toJSON (RichInfoAssocList l) =
-    object
-      [ "fields" .= l,
-        "version" .= (0 :: Int)
-      ]
+instance S.ToSchema RichInfoAssocList where
+  schema =
+    S.object "RichInfoAssocList" $
+      fmap snd $ (,)
+          -- TODO(sandy): check the version is equal to 0
```

Comment:
@pcapriotti The `FromJSON` instance immediately after this checks to ensure the version is 0, failing otherwise. Is this a check I can replicate in schema-profunctor?

Furthermore, this `version` doesn't actually end up in the datatype. Is there a better way than this ugly `fmap snd $ (,) ...` thing I'm doing here?

Reply:
Yes to both.

To add a check, you can use `withParser`, with a function that performs the check and fails with `fail` in the `Aeson.Parser` monad. `withParser` allows you to change the type, so that answers your second question as well.

So for example you can have something like:
```haskell
withParser ((,) <$> field1 <*> field2) $ \(v, xs) -> checkVersion v $> xs
```

Also, there's no need to use `dimap` here. `const 0 .= schema` should work fine (the annotation can probably be removed once you add `withParser` in).